qiskit/transpiler/passes/basis/basis_translator.py

```diff
@@ -391,3 +440,20 @@ def _compose_transforms(basis_transforms, source_basis, source_dag):
                 )
 
     return mapped_instrs
+
+
+def _compute_incomplete_basis(target):
+    incomplete_basis_gates = []
+    size_dict = defaultdict(lambda: 0)
+    size_dict[1] = target.num_qubits
+    for qarg in target.qargs:
+        if len(qarg) == 1:
+            continue
+        size_dict[len(qarg)] += 1
+    for inst, qargs in target.items():
+        qarg_sample = next(iter(qargs))
+        if qarg_sample is None:
+            continue
+        if len(qargs) != size_dict[len(qarg_sample)]:
+            incomplete_basis_gates.append(inst)
```

Comment:
I don't think this actually tests if the gate applies to all qubits, right?  If I have a target with no coupling at all between qubits 0 and 2 and a single 2q gate, then this test will appear true, because the `qargs` will be exactly the set of operands to this gate, so `size_dict[2]` will be the same as `len(qargs)`, even though there's not full connectivity (which I what I think the previous iteration of this function was testing).

Also, minor sidenote: do targets need to specify symmetric gates like `cz` with qargs `(a, b)` and `(b, a)`, or only one of the pair?

Reply:
Ah, yeah my bad I made a copy paste error this should be `len(qargs)` not `len(qarg_sample)`. 

Right the earlier iteration was doing the wrong thing. We want this to return a list of instructions which do not apply globally in the target, for 1q this is on all qubits, for 2q this is on all edges in the coupling graph, >3q on all valid qargs in the target of that size. The previous iteration was incorrect because it was always checking the full permutation which was only true on an all to all coupling map (it was only correct in the 1q case).

> Also, minor sidenote: do targets need to specify symmetric gates like cz with qargs (a, b) and (b, a), or only one of the pair?

They need to be specified bidirectionally (so `(a, b)` and `(b, a)`) the way the target is implemented now. Otherwise the target will treat it like you can only run cz on `(a, b)` or `(b, a)` (whichever you put in the target).

Reply:
With the new explanation, I think what you've got written here is correct.  That also makes way more sense to me - I was stuck still thinking about the old version of the function as what we were trying to achieve.

Reply:
Taking another look at it I actually I think I did this correctly, I'm trying to get the size of the qargs tuple to look up the number of valid options in the target. So for example if we have a 3q bidirectional linear coupling map: `[(0, 1), (1, 0), (1, 2), (2, 1)]` and have cx defined on all edges, but ecr only on reverse (so `(1, 0)` and (2, 1)`). The check here should be:

For ecr:
```
size_dict = {1: 3, 2: 4}

len([(1, 0), (2, 1)]) != len(size_dict[2])
```
which will add ecr to the incomplete list.

While for cx it'll be:

```
len([(0, 1), (1, 0), (1, 2), (2, 1)] != size_dict[2]
```
which means we don't add cx to the incomplete list because it's on all edges in the coupling graph.

Reply:
Yeah, I agree - I thought it was wrong because I thought we were shooting for the old, incorrect behaviour.

Reply:
I think we do at least until we can make gate direction target aware. The other thing is if we wanted to rely on gate direction for this we'd need to differentiate between unidirectional gates vs bidirectional gates in the target (at least for the `compute_non_global_operation_names()` method which it doesn't do right now.

Reply:
I don't think gate direction needs to be target aware (yet), but I do think the `BasisTranslator` should not (maybe not yet) be gate direction aware. Otherwise, e.g. if we have the equivalent of a non-symmetric coupling map (with some non-global gates), the `BasisTranslator` would raise before we get a chance to run `GateDirection`.

Reply:
The problem here is that in the case where you have a target that's like what's in `FakeBackendV2` where cx is on `(0, 1)` and ecr is on `(1, 0)` if you treat each 2q gate as symmetric (or a weak edge in a directed graph) if you have a circuit with a 2q gate on `(1, 0)` outside the basis you'll almost always end up with `cx(1, 0)` in the output circuit. This is also why gate direction being target aware is a requirement because the coupling map object that gets used is of the global edges independent of which gates so gate direction will think that's valid unless it's made target aware. That being said even if you do this and make gate direction target aware (I have a commit on this PR branch locally doing it) this still ends up with a less than ideal result because you'll end up with `cx(0, 1)` with the `h` gates around it where `ecr(1, 0)` is a valid and less gates